Add options to embolden and transform font outlines to simulate bold and italic typefaces.

bruvzg opened the pull request:
- Adds options to embolden and transform font outlines to simulate bold and italic typefaces.
- Adds bold, bold italics, and italics modified font versions to the default theme (for `RichTextLabel` style). With `1.2` embolden and `Transform2D(1.0, 0.4, 0.0, 1.0, 0.0, 0.0)` transform.

```
Normal
[b]Bold[/b]
[i]Italic[/i]
```

<img width="109" alt="Screenshot 2022-03-11 at 09 28 55" src="https://user-images.githubusercontent.com/7645683/157822716-f79a512b-ba7c-48e0-9fc5-e2ff378a0cd8.png">

Fixes #58988

bruvzg committed:
Add options to embolden and transform font outlines to simulate bold and italic typefaces.

bruvzg commented:
https://user-images.githubusercontent.com/7645683/157846669-2759abfb-b012-4f81-b03b-5740a98ade46.mov

akien-mga commented:
Playing a bit with dynamic updates to `slant` and `embolden`:
https://user-images.githubusercontent.com/4701338/157852110-4138f29f-31b4-4662-a25a-078f53529702.mp4

It works well but seems to be fairly expensive, the above drops from 1670 FPS with just the RichTextEffects, to 20 FPS with both `slant` and `embolden` set dynamically in `_process`. Or down to 2-3 FPS if set in `_physics_process`.

If I double the font size from 48 to 96px, I get 5 FPS in `_process`.

Using this code:
```
var fd = get_theme_font("normal_font").get_data(0)

func _process(delta):
	fd.slant = sin(Time.get_ticks_msec() / 1000.0)
	fd.embolden = clamp(cos(Time.get_ticks_msec() / 500.0), 0.2, 0.8)
```

bruvzg commented:
Change `slant` to full Transform2D, now it's possible to do more stuff:

https://user-images.githubusercontent.com/7645683/157855628-bdfd762c-7f55-4da7-a989-3893e5a3c88f.mov

Original `slant` is equivalent to `Transform2D(1.0, 0.4, 0.0, 1.0, 0.0, 0.0)`

akien-mga commented:
> Changed `slant` to full Transform2D, now it's possible to do more stuff:

Heh I was suggesting this mostly in jest, I don't know if there's a real need for it.

It's nice to have the flexibility but at the same time I fear it might not be easy for users to edit the Transform2D directly if what they want is a simple slanting effect.

Maybe mention in the `transform` documentation that for italics-like slanting users should set the `yx` component?

bruvzg commented:
> Maybe mention in the transform documentation that for italics-like slanting users should set the yx component?

Added `slant` an example to the documentation.

Also changed max. font texture cache size to `1024x1024` (`2048x2048` for MSDF, it's slow anyway and usually have large size glyphs rendered), it should be faster with larger font.

akien-mga commented:
> Also changed max. font texture cache size to `1024x1024` (`2048x2048` for MSDF, it's slow anyway and usually have large size glyphs rendered), it should be faster with larger font.

It doesn't seem to make much difference in my test project, still ~40 FPS with 64px font and 10 FPS with 65px font when changing `embolden` and `transform` every other `_process` call.

bruvzg commented:
> 64px font and 10 FPS with 65px

This should be transition from `512` to `1024`, so it's not changed.

akien-mga commented:
Thanks!

markdibarry commented:
It seems this causes all font outlines to be anti-aliased. Is that intentional?
